### ISSUE TITLE
datapipe Error in SCurve test

### ISSUE THREAD

**Issue by thweiler** [psi46/pxar#357]
Hi
In the Scurve test I get following error:
[10:49:08.275]    ERROR: <datapipe.cc/CheckEventValidity:L420> Number of ROCs (7) != Token Chain Length (8)
[10:49:08.275]    ERROR: <datapipe.cc/CheckEventID:L154>    Event ID mismatch:  local ID (25) !=  TBM ID (26)
[10:49:08.275]  WARNING: ROC 0: Readback start marker after 15 readouts!
[10:49:08.275]  WARNING: ROC 1: Readback start marker after 15 readouts!

This always happens in the same step of the test, after the dacScanStep 80 ... 89 meaasage

[10:48:25.157]     INFO: Fetched DAQ statistics. Counters are being reset now.
[10:48:25.157]     INFO:   dacScan step  from 80 .. 89

the data pipe errors occur and stay until the Scurve test is finished, in the following test
(Trim, phOptimization and GainPedestal) there are no errors. 

pXar actual version with fw4.0
Test devices: full modules 3 with TBM08BB, 5 with TBM08C all with digv21respin
Two different test setup (in different labs)

Does anybody see the same behaviour?

cheers
Thomas

**Comment by ursl** [comment]
Hi Thomas, all,

I also observe similar behavior. However, for me it is not as
clear-cut as for you. I have cases where the problems start with the
70 .. 79 range, and most of the time the problems continue for a few
ranges.

I also (most of the times) do not have problems afterwards. The big
(only) difference between the relevant trim code and the s-curves is
ntrig.

The corollary of this is to replace the splitting into dac ranges with
a split into smaller ntrig and assemble enough statistics in that way.
An alternative might be to "re-initialize" "something".

I'll try that.

Cheers,
--U.

**Comment by simonspa** [comment]
I think in general it might be useful to replace the calls to `fApi->getStatistics()` in PixTest with something that actually reads them and print them to screen:

```
pxar::statistics mystats = fApi->getStatistics();
mystats.dump()
```

so we see the full list of errors available. It of course means a bit more screen output, check it out yourself to see if it suits.

Also fro this problem it might be interesting. Is that the noisy region?

**Comment by ursl** [comment]
Hi Simon,

will do (maybe only when problems were encountered).

This is a VCAL scan for the scurves, there is no noisy region per se.

Cheers,
--U.

**Comment by simonspa** [comment]
Okay, then at least this shouldn't be the problem...

Make sure to store the statistics before checking for errors, you can only read them once. There is a member `pxar::statistics::errors()` returning the sum of all problems which is probably what you want to check before printing.

**Comment by drberry85** [comment]
I highly suspect this issue is related to issue #325...

Doug

**Comment by simonspa** [comment]
hut @thweiler claims that the errors continue to be thrown, while #325 is just one single garbled event, right?

**Comment by drberry85** [comment]
Simon,

That's what I saw originally. When I initially opened the issue on FW 3.6, the error was most easily seen in the timing test with 2-3 million triggers. After FW 4.0, this error went away. However, the very same problem can still be seen in the SCurves measurement. If you look at my logs, you can see I run an SCurve measurement on three different DTBs, and each time the decoding errors start after step 89.
https://drberry.web.cern.ch/drberry/DecodingIssues/dtb27.log
https://drberry.web.cern.ch/drberry/DecodingIssues/dtb38.log
https://drberry.web.cern.ch/drberry/DecodingIssues/dtb82.log

The only problem is I stopped the VCAL scan at 100 instead of going all the way to 255 (because I'm impatient). But if you look at the logs to can see the raw event dumps of the errors during the SCurves measurement:
```
[15:26:57.983]     INFO: Fetched DAQ statistics. Counters are being reset now.
[15:26:57.989]     INFO:   dacScan step  from 94 .. 95
[15:26:57.990]    DEBUG: <PixTest.cc/dacScan:L1564>       attempt #0
[15:27:27.004]    ERROR: <datapipe.cc/CheckEventValidity:L423> Number of ROCs (4) != Token Chain Length (8)
[15:27:27.004]    DEBUG: <datapipe.cc/DecodeDeser400:L316> ====== 2 ====== a067 8041 4878 6dc 2668 4878 6dc 266a 4878 4878 10 0 1020 20 2118 8 420 10 140 40 102 0 
[15:27:27.004]    DEBUG: <datapipe.cc/DecodeDeser400:L317> Event Buffer:
[15:27:27.004]    DEBUG: <datapipe.cc/DecodeDeser400:L318> ====== 0 ====== a063 8041 4878 6dc 2668 4879 4878 4879 6dc 2683 4878 4879 6dc 2688 4878 6dc 26a2 4879 6dc 2683 e022 c080 
[15:27:27.004]    DEBUG: <datapipe.cc/DecodeDeser400:L318> ====== 0 ====== a064 80b1 4878 6dc 2669 4878 4878 4878 6dc 2683 4878 4878 6dc 2689 4878 6dc 26a2 4878 6dc 2683 e022 c080 
[15:27:27.004]    DEBUG: <datapipe.cc/DecodeDeser400:L318> ====== 0 ====== a065 80c1 4878 6dc 2669 4878 6dc 266c 4878 4878 6dc 2685 4878 4878 6dc 2686 4878 6dc 26a1 4878 6dc 2683 e022 c080 
[15:27:27.004]    DEBUG: <datapipe.cc/DecodeDeser400:L318> ====== 0 ====== a066 8001 4878 6dc 2668 4878 6dc 266c 4878 4878 6dc 2683 4878 4878 6dc 2687 4878 6dc 26a1 4879 e022 c080 
[15:27:27.004]    DEBUG: <datapipe.cc/DecodeDeser400:L318> ====== 2 ====== a067 8041 4878 6dc 2668 4878 6dc 266a 4878 4878 10 0 1020 20 2118 8 420 10 140 40 102 0 
[15:27:27.004]    DEBUG: <datapipe.cc/DecodeDeser400:L322> ====== 0 ====== a068 80b1 4878 6dc 2668 4878 6dc 266a 4878 4878 6dc 2685 4878 4878 6dc 2687 4878 6dc 26a4 4879 e022 c080 
[15:27:27.004]    DEBUG: <datapipe.cc/DecodeDeser400:L322> ====== 0 ====== a069 80c1 4879 6dc 2668 4879 6dc 266b 4879 4879 6dc 2683 4879 4879 6dc 2688 4879 6dc 26a1 4879 6dc 2683 e022 c080 
[15:27:27.004]    DEBUG: <datapipe.cc/DecodeDeser400:L322> ====== 0 ====== a06a 8001 4878 6dc 2668 4878 6dc 266c 4878 4878 6dc 2683 4878 4878 6dc 2688 4878 6dc 26a1 4878 6dc 2683 e022 c080 
[15:27:27.004]    DEBUG: <datapipe.cc/DecodeDeser400:L322> ====== 0 ====== a06b 8041 4879 6dc 2668 4879 4879 4879 6dc 2685 4879 4879 4879 6dc 26a3 4879 6dc 2683 e022 c080 
[15:27:27.004]    DEBUG: <datapipe.cc/DecodeDeser400:L322> ====== 0 ====== a06c 80b1 4879 6dc 2668 4879 4879 4879 6dc 2683 4879 4879 6dc 2687 4879 6dc 26a2 4879 6dc 2681 e022 c080 
[15:27:27.004]  WARNING: ROC 4: Readback start marker after 15 readouts!
[15:27:27.004]  WARNING: ROC 5: Readback start marker after 15 readouts!
[15:27:27.004]  WARNING: ROC 6: Readback start marker after 15 readouts!
[15:27:27.004]  WARNING: ROC 7: Readback start marker after 15 readouts!
[15:27:30.146]     INFO: Test took 32156ms.
```
Here it is clear that event 67 lost an entire TBM port for dac step 94-95. Or more likely the data just got garbled. You can see port 0 is fine: ``` a067 8041 4878 6dc 2668 4878 6dc 266a 4878 4878 ```
but port 1 is toast: ```10 0 1020 20 2118 8 420 10 140 40 102 0 ```.
I would note though that I am running with different settings than @thweiler, but if he sends me his setting (dacname, start, stop, step). I am sure I can reproduce the problem.

Doug

**Comment by thweiler** [comment]
Hi Doug
I am just running a fulltest via 'cat main/mtest | bin/pXar -d ...'
see PixTestScurve::fulltest():
  fParNtrig = 50; 
  fParDac = "Vcal"; 
  fParDacLo = 0; 
  fParDacHi = 149;
  fParDacsPerStep = 10;   

cheers
Thomas

**Comment by simonspa** [comment]
Just to understand correctly, @drberry85: it still is only a single even (in this case `0x67`)?
How is that for you, @thweiler, you also just see one error being thrown or thousands?

**Comment by thweiler** [comment]
Hi Simon
It looks like the attched output, the error comes at dacScan step from 80..89
and stays there until the end of the Scurve test, the following are without errors.

[08:42:45.314]     INFO: Fetched DAQ statistics. Counters are being reset now.
[08:42:45.314]     INFO:   dacScan step  from 80 .. 89
[08:43:30.083]    ERROR: <datapipe.cc/CheckEventValidity:L420> Number of ROCs (9) != Token Chain Length (8)
[08:43:30.083]    ERROR: <datapipe.cc/CheckEventID:L154>    Event ID mismatch:  local ID (224) !=  TBM ID (225)
[08:43:30.083]  WARNING: ROC 3: Readback start marker after 15 readouts!
[08:43:30.083]  WARNING: ROC 0: Readback start marker after 15 readouts!
[08:43:30.083]  WARNING: ROC 1: Readback start marker after 15 readouts!
[08:43:30.083]  WARNING: ROC 2: Readback start marker after 15 readouts!
[08:43:30.083]  WARNING: ROC 4: Readback start marker after 15 readouts!
[08:43:30.083]  WARNING: ROC 5: Readback start marker after 15 readouts!
[08:43:30.083]  WARNING: ROC 6: Readback start marker after 15 readouts!
[08:43:30.083]  WARNING: ROC 7: Readback start marker after 15 readouts!
[08:43:30.264]     INFO: Test took 44950ms.
[08:43:30.370]     INFO: Fetched DAQ statistics. Counters are being reset now.
[08:43:30.373]     INFO:   dacScan step  from 90 .. 99
[08:44:15.931]    ERROR: <datapipe.cc/CheckEventValidity:L420> Number of ROCs (0) != Token Chain Length (8)
[08:44:15.931]    ERROR: <datapipe.cc/CheckEventID:L154>    Event ID mismatch:  local ID (99) !=  TBM ID (100)
[08:44:15.931]  WARNING: ROC 3: Readback start marker after 14 readouts!
[08:44:15.931]  WARNING: ROC 0: Readback start marker after 14 readouts!
[08:44:15.931]  WARNING: ROC 1: Readback start marker after 14 readouts!
[08:44:15.931]  WARNING: ROC 2: Readback start marker after 14 readouts!
[08:44:15.931]  WARNING: ROC 4: Readback start marker after 14 readouts!
[08:44:15.931]  WARNING: ROC 5: Readback start marker after 14 readouts!
[08:44:15.931]  WARNING: ROC 6: Readback start marker after 14 readouts!
[08:44:15.931]  WARNING: ROC 7: Readback start marker after 14 readouts!
[08:44:19.948]     INFO: Test took 49575ms.
[08:44:20.185]     INFO: Fetched DAQ statistics. Counters are being reset now.
[08:44:20.196]     INFO:   dacScan step  from 100 .. 109
[08:45:04.287]  WARNING: ROC 6: Readback start marker after 4 readouts!
[08:45:04.287]    ERROR: <datapipe.cc/CheckEventValidity:L420> Number of ROCs (7) != Token Chain Length (8)
[08:45:04.287]    ERROR: <datapipe.cc/CheckEventID:L154>    Event ID mismatch:  local ID (36) !=  TBM ID (37)
[08:45:04.287]  WARNING: ROC 3: Readback start marker after 15 readouts!
[08:45:04.287]  WARNING: ROC 0: Readback start marker after 15 readouts!
.
.
.
[08:49:11.061]     INFO:   dacScan step  from 140 .. 149
[08:49:51.739]  WARNING: ROC 2: Readback start marker after 3 readouts!
[08:49:51.739]    ERROR: <datapipe.cc/CheckEventValidity:L420> Number of ROCs (3) != Token Chain Length (8)
[08:49:51.739]  WARNING: ROC 3: Readback start marker after 15 readouts!
[08:49:51.739]  WARNING: ROC 2: Readback start marker after 13 readouts!
[08:49:51.739]  WARNING: ROC 4: Readback start marker after 15 readouts!
[08:49:51.739]  WARNING: ROC 5: Readback start marker after 15 readouts!
[08:49:51.740]  WARNING: ROC 6: Readback start marker after 15 readouts!
[08:49:51.740]  WARNING: ROC 7: Readback start marker after 15 readouts!
[08:50:22.338]    ERROR: <datapipe.cc/CheckEventValidity:L420> Number of ROCs (9) != Token Chain Length (8)
[08:50:22.338]    ERROR: <datapipe.cc/CheckEventID:L154>    Event ID mismatch:  local ID (49) !=  TBM ID (50)
[08:50:22.338]  WARNING: ROC 3: Readback start marker after 15 readouts!
[08:50:22.338]  WARNING: ROC 0: Readback start marker after 15 readouts!
[08:50:22.338]  WARNING: ROC 1: Readback start marker after 15 readouts!
[08:50:22.338]  WARNING: ROC 2: Readback start marker after 15 readouts!
[08:50:22.338]  WARNING: ROC 4: Readback start marker after 15 readouts!
[08:50:22.338]  WARNING: ROC 5: Readback start marker after 15 readouts!
[08:50:22.338]  WARNING: ROC 6: Readback start marker after 15 readouts!
[08:50:22.338]  WARNING: ROC 7: Readback start marker after 15 readouts!
[08:50:29.580]     INFO: Test took 78519ms.
[08:50:30.145]     INFO: Fetched DAQ statistics. Counters are being reset now.
[08:50:52.027]     INFO: PixTestScurves::scurves() done 
[08:50:52.027]     INFO: Vcal mean: 109.14 100.52 106.03 112.36 101.16 104.89 104.60  90.96 104.92 100.74  98.05 111.58  99.31 127.62 102.44 100.92 
[08:50:52.027]     INFO: Vcal RMS:    5.15   6.86   5.49   4.97   5.98   6.04   5.54   5.50   5.09   5.43   5.30   4.53   5.02   7.20   6.65   5.74 
[08:50:52.027]     INFO: PixTestScurves::fullTest() done, duration: 1239 seconds
[08:50:52.209]     INFO: ######################################################################
[08:50:52.209]     INFO: PixTestTrim::doTest()
[08:50:52.209]     INFO: ######################################################################
[08:50:52.210]     INFO:    ----------------------------------------------------------------------
[08:50:52.210]     INFO:    PixTestTrim::trimTest() ntrig = 10, vcal = 35
[08:50:52.210]     INFO:    ----------------------------------------------------------------------
[08:50:52.297]     INFO: ---> VthrComp thr map (minimal VthrComp)
[08:50:52.297]     INFO: ---> dac: vthrcomp name: TrimThr0 ntrig: 5 dacrange: 0 .. 159 (20) hits flags = 16 (plus default)
[08:50:52.306]     INFO:   dacScan step  from 0 .. 19
[08:51:07.755]     INFO: Test took 15449ms.
[08:51:07.778]     INFO: Fetched DAQ statistics. Counters are being reset now.
[08:51:07.778]     INFO:   dacScan step  from 20 .. 39
[08:51:23.136]     INFO: Test took 15358ms.
[08:51:23.164]     INFO: Fetched DAQ statistics. Counters are being reset now.
[08:51:23.164]     INFO:   dacScan step  from 40 .. 59
[08:51:38.428]     INFO: Test took 15264ms.
[08:51:38.455]     INFO: Fetched DAQ statistics. Counters are being reset now.
[08:51:38.455]     INFO:   dacScan step  from 60 .. 79





Cheers
Thomas

**Comment by simonspa** [comment]
Since after every ` [08:42:45.314] INFO: dacScan step from 80 .. 89` a new test is started (resulting in a DAQ clear command) the error occurs once or twice during every test.
This doesn't really help me to understand it better... :(

**Comment by werdmann** [comment]
For the records, some observations with TBM09c modules.
 I observe the same problem, it is obviously related with issue #325.

Listing one event before and after each event with an error showed interesting patterns. I have no explanation for this yet but I find it hard to explain this with problems decoding the 400 MHz data.

1) The "bad" event id in the tbm header is the same for a series of errors. In this case it is 245, but I have seen other numbers in other runs, too.
2) the event before the 'bad' event is often lacking roc headers. Instead it has a hit like 0314 248x. It looks like valid data, but I don't think this hit is expected here. All preceding and all following events don't have it (except for the bad event).
3) the XOR pattern in the bad event's roc headers is x70, which never occurs in any good event.
  All other events in the entire test seem to have x18 or x38. The events before and after the bad event both have x18. To my knowledge, the testboard is not expected to update the the phase measurement from one event to the next and then back again immediately.
4) the event number difference from the first the second error is  3990980-1910980 = 2080000
The next difference is 6070980-3990980 =  2080000
The third error comes after 8149853-6070980 = 2078873 and after that the pattern is a bit diluted.


Below comes the summary of the errors from the scurve run  (only channel 3):

event id mismatch in event 1910980  :  got 245   expectecd  196
   1910979:  a0c3 8041 4188 4189 0315 248f e022 c000
   1910980:  a0f5 80c0 4708 0315 2480 4708 0315 2482
   1910981:  a0c5 80c1 4188 4188 4188 4189 e022 c000

the corresponding pxar messages are
[11:14:09.159]     INFO:   dacScan step  from 0 .. 9
[11:14:46.478]    ERROR: <datapipe.cc/CheckEventValidity:L437> Channel 3 Number of ROCs (2) != Token Chain Length (4)
[11:14:46.479]    ERROR: <datapipe.cc/CheckEventID:L420> Channel 3 Event ID mismatch:  local ID (196) !=  TBM ID (245)
[11:14:46.479]    ERROR: <datapipe.cc/CheckEventValidity:L437> Channel 3 Number of ROCs (2) != Token Chain Length (4)
[11:14:46.479]    ERROR: <datapipe.cc/CheckEventID:L420> Channel 3 Event ID mismatch:  local ID (246) !=  TBM ID (197)

more errors

event id mismatch in event 3990980  :  got 245   expectecd  196
   3990979:  a0c3 8041 4188 4189 0315 248f e022 c000
   3990980:  a0f5 80c0 4708 0315 2480 4708 0315 2482
   3990981:  a0c5 80c1 4188 4188 4188 4189 e022 c000

event id mismatch in event 6070980  :  got 245   expectecd  196
   6070979:  a0c3 8041 4188 4189 0315 248f e022 c000
   6070980:  a0f5 80c0 4708 0315 2480 4708 0315 2482
   6070981:  a0c5 80c1 4188 4188 4188 4189 e022 c000

event id mismatch in event 8149853  :  got 245   expectecd  93
event id mismatch in event 8149854  :  got 95   expectecd  94
   8149852:  a05c 80b1 4188 4188 4189 4189 0315 248f e022 c000
   8149853:  a0f5 80c0 4708 0315 2480 4708 0315 2482 4188 4188 418b 4189 e022 c000
   8149854:  a05f 8041 418b 418b 4189 4189 e022 c000

event id mismatch in event 9134637  :  got 47   expectecd  46
event id mismatch in event 9134638  :  got 48   expectecd  47
   9134636:  a02d 80c1 4388 4388 4389 4389 e022 c000
   9134637:  a02f 8041 418a 418a 4189 4189 e022 c000
   9134638:  a030 80b1 4189 4189 4189 4189 e022 c000

event id mismatch in event 10056356  :  got 245   expectecd  166
  10056355:  a0a5 0315 248f e022 c000
  10056356:  a0f5 80c0 4708 0315 2480 4708 0315 2482 4388 4388 4389 054a 248d e022 c000
  10056357:  a0a7 8041 4388 4388 4388 4389 054a 2489 e022 c000

event id mismatch in event 11617193  :  got 245   expectecd  171
  11617192:  a0aa 8001 4389 4389 0309 2a81 4389 0309 2aa1 4388 0309 2aa1 0fff 2fff e022 c000 0315 248f e022 c000
  11617193:  a0f5 80c0 4708 0315 2480 4708 0315 2482
  11617194:  a0ac 80b1 4388 4388 0309 2a81 4389 4388 0309 2aa1 e022 c000

event id mismatch in event 12431394  :  got 245   expectecd  36
  12431393:  a023 8041 4388 06e1 208d 4389 06e1 208e 0315 248f e022 c000
  12431394:  a0f5 80c0 4708 0315 2480 4708 0315 2482 208d 4388 06e1 208d 4388 06e1 20a4 4389 06e1 20a3 e022 c000
  12431395:  a025 80c1 4388 06e1 208c 4388 06e1 208e 4388 06e1 20a3 4389 06e1 20a3 e022 c000

**Comment by werdmann** [comment]
Addendum to the previous post.  That scurve run was done with Ntrig=50. No error is observed for ntrig=10 while for ntrig=53 the position of the error moves from 2080000 to 2204800. 
In both cases this is Ntrig x 41600.  That factor happens to be the product of the number of pixels per roc and the number of dacs per step.
Changing the latter to -1 after discussion with Urs produced the described errors in distance very close to multiples 12500000 words (not events):[0.0, 1.0, 3.99999984, 8.99999984, 9.99999984, 10.99999984, 11.99999984]
I can't help noticing that this is the DTB buffer size divided by the number of DAQ channels and I finally stop assuming that this has anything to do with the TBM phase settings.

**Comment by simonspa** [comment]
Hi @werdmann, this is quite an interesting insight and matches more or less with what I last discussed with Beat at the workshop. Starting from the other side of the readout chain:

* it's unlikely to be a USB problem since it really is restricted to a single event as you have shown
* it's also unlikely to be a problem with the trigger loops on the DTB. The reason for this is that interrupting the loop and downloading the data happens at 80% fill level of the allocated memory, not when it's full. If we somehow screw up the data at this point I would rather expect the problematic event to be around 80% of 12500000 words
* leaves the ring buffer implementation of the DTB. 

When interrupting a loop due to full memory (80%) we read all the data from the DTB RAM. After reading everything we currently don't reset the DAQ buffers  but just continue writing data. This results in a pointer wrap over at the end of the allocated memory for every channel (just as expected from a ring buffer).
This implementation from Beat in principle allows endless data acquisition in case of using a trigger source which does not occupy the NIOS core (such as external trigger for example) since data can be read continuously while new data ist written to the RAM by the deserializers.

If the problems really always occur when jumping from the end of the allocated memory to the beginning I would suspect the ring buffer mechanism to be the culprit - which is something Beat should then probably look into...

**Comment by simonspa** [comment]
I am not able to reproduce this using a single ROC and the DESER160. My setup is the following:

* single ROC (no sensor) psi46dig
* RAM allocation: `#define DTB_SOURCE_BUFFER_SIZE 12288`
* 9 pixels active responding to VCal pulses
* Sending 450 triggers leads to collected data of 8550 words
* Ring buffer should wrap over after 647 triggers (=12293 words)

the script does the following (after enabling pixels):
```
# start DAQ
daqStart
# trigger 450 times with 5000 BC spacing
daqTrigger 450 5000
# get all events, fully decoded
daqGetEventBuffer
# read the statistics on errors etc
getStatistics

# now, lets trigger again - enough space in the buffer, but ring buffer wrap over ahead:
daqTrigger 450 5000
daqGetEventBuffer
getStatistics

```
but I don't see any errors, just perfectly fine data with all pixels in as expected.

**Comment by simonspa** [comment]
Same test, different setup:

* Full preseries Module (TBM08b, 16 ROCs, DESY Module D4020)
* RAM allocation: `#define DTB_SOURCE_BUFFER_SIZE 32768`
* which means we have 16384 word per channel allocated.
* 9 pixels active on every ROC responding to VCal pulses
* Sending 1 trigger leads to collected data of 1555 words per channel
* Ring buffer should wrap over after 106 triggers (=16430 words)

the script does the following (after enabling pixels):

```
# start DAQ
daqStart
# trigger 100 times with 5000 BC spacing
daqTrigger 100 5000
# get all events, fully decoded
daqGetEventBuffer
# read the statistics on errors etc
getStatistics

# now, lets trigger again - enough space in the buffer, but ring buffer wrap over ahead:
daqTrigger 100 5000
daqGetEventBuffer
getStatistics
```

but I don't see any errors, just perfectly fine data with all pixels in as expected.

**Comment by simonspa** [comment]
It does not seem to depend on the time between triggers, reducing this works also fine until I'm cutting in the event. There the error pattern looks different, we are just losing the last ROC(s):

```
19:37:58.258] DEBUGPIPES: a147 8041 4c18 021c 2765 4c18 021c 274b 4c18 021c 2764 4c18 021c 274b 4c19 021f 2fff 0fff e0a2 c000

[19:38:18.730]     INFO:   TBM errors: 		           0
[19:38:18.730]     INFO:   ROC errors: 		           1572
[19:38:18.730]     INFO: 	 missing ROC header(s):    1568

```

I reduced the number of pixels active to one per ROC which let me go down to a trigger spacing of roughly 200 BC with out the errors described above.

**Comment by werdmann** [comment]
Hi Simon,

    is there anything special to the way the deser400 writes to that 
memory? Their packaging looks like deserializer output, only the payload 
doesn't make sense.  They are fairly reproducable and don't seem to be 
leftovers from earlier datataking still lying around. I don't know 
whether this makes sense.

Wolfram

**Comment by werdmann** [comment]
I have tried to reproduce the problem outside of the scurve test, but  I have not managed to see it using combinations of daqTrigger + daqGetRawEventBuffer.
The following code on the other hand produces the error consistently in my setup:

        fApi->_dut->testAllPixels(true);
        fApi->_dut->maskAllPixels(false);
        try{
            fApi->getEfficiencyVsDAC("CalDel", 1, ndac, FLAG_FORCE_MASKED, ntrig);
            cout <<  fApi->getStatistics().errors_pixel() << " errors " << endl;
        }catch(pxar::DataMissingEvent){  out << "DataMissingEvent caught\n";  }

Whenever ndac x ntrig x 4160 x 8 is larger than 12500000, errors of the type described above appear. This code is in the current head of the cmdline and can be invoked by typing e.g.
looptest 376 1  (the first number is the ntrig, the second ndac).
This produces the error, while looptest 375 1  does not (unless you have other problems, too).

I'm using a tbm09c module, my guess would be that for a tbm08c module the equation reads 
ndac x ntrig x 4160 x 12 > 25000000 because it uses only two channels and the number of words for an empty readout is 12 instead of 8.

The error does occur near the 12500000th word, not at the position where the loop is interrupted, so I don't understand why this is different from just repeatedly sending triggers and reading out. If somebody knows what the fundamental difference is, it might be hint towards whats going wrong.

**Comment by werdmann** [comment]
More blogging on this issue:
A similar effect can be seen in logfiles of xray tests I received and it was possible to reproduce it using similar code with a sequence of daqStart, daqTriggerLoop,daqGetEventBuffer, daqStop.
Again, the positions of the error in the datastreams support the hypothesis, that the wrap-around of the DTB buffer is  involved. A modified version of hal::daqTriggerLoop which injects a dtb Daq_Open (resetting the internal memory pointers) removes the errors in these tests. 
Unfortunately doing the same thing for scurve tests would require modifying the nios code and is a bit more difficult to test.

If you would like to try it in a real x-ray test, the modified hal.cc  can be found at 
/afs/cern.ch/user/w/werdmann/public/hal.cc

**Comment by simonspa** [comment]
Dear @thweiler, this issue should now be resolved with FW4.4 and pxarCore 2.5.
Please check this and open a new ticket if the problem persists.
